src/v/cluster/log_eviction_stm.cc

```diff
@@ -9,104 +9,345 @@
+        vlog(
+          _logger.debug,
+          "Reaping data before offset: {} - for ntp: {}",
+          nearest_segment_boundary,
+          _c->ntp());
+        co_await _raft->write_snapshot(
+          raft::write_snapshot_cfg(nearest_segment_boundary, iobuf()));
+    } else {
         vlog(
           _logger.trace,
-          "Can only evict up to offset: {}, ntp: {}",
-          max_collectible_offset,
+          "Cannot prefix truncate due to ongoing processing on data "
+          "below the requested new start offset {} for ntp {}, "
+          "retrying... ",
+          nearest_segment_boundary,
           _raft->ntp());
-        last_evicted = max_collectible_offset;
+    }
+}
+
+model::offset log_eviction_stm::effective_start_offset() const {
+    /// The start offset is the max of either the last snapshot index or the
+    /// most recent delete records eviciton offset. This is because even after
+    /// bootstrap the larger of the two will reflect the most recent event that
+    /// has occurred, and will be the correct start offset.
+    ///
+    /// NOTE: Cannot replace last_snapshot_index with _storage_eviction_offset
+    /// as this is the requested eviction offset and its not persisted anywhere.
+    /// In the event this is set but a crash occurred before write_snapshot was
+    /// called (occurs in background) it would appear that the start offset was
+    /// incremented then returned to a previous value.
+    return model::next_offset(
+      std::max(_raft->last_snapshot_index(), _delete_records_eviction_offset));
+}
+
+ss::future<std::error_code> log_eviction_stm::truncate(
+  model::offset kafka_truncate_offset,
+  ss::lowres_clock::time_point deadline,
+  std::optional<std::reference_wrapper<ss::abort_source>> as) {
+    /// Create the special prefix_truncate batch, it is a model::record_batch
+    /// with exactly one record within it, the point at which to truncate
+    storage::record_batch_builder builder(
+      model::record_batch_type::prefix_truncate, model::offset(0));
+    /// Everthing below the requested offset should be truncated, requested
+    /// offset itself will be the new low_watermark (readable)
+    const model::offset as_log_offset
+      = _raft->get_offset_translator_state()->to_log_offset(
+        kafka_truncate_offset);
+    auto key = serde::to_iobuf(as_log_offset - model::offset{1});
+    builder.add_raw_kv(std::move(key), iobuf());
+    auto batch = std::move(builder).build();
+    /// After command replication all that can be guaranteed is that the command
+    /// was replicated
+    vlog(
+      _logger.info,
+      "Replicating prefix_truncate command, truncate_offset: {} current start "
+      "offset: {}, current last offset: {}",
+      kafka_truncate_offset,
+      _raft->last_snapshot_index(),
+      _raft->last_visible_index());
+    auto result = co_await replicate_commands(std::move(batch), deadline, as);
+    if (result) {
+        co_return result;
+    }
```

Comment:
This logic is pretty tricky, we ignore an error and move to waiting, maybe it would be good to place a comment here and log the result with debug log ?